src/XRP/network-clients/web-socket-network-client.ts

```diff
@@ -0,0 +1,169 @@
+import WebSocket = require('ws')
+import { XrpError, XrpErrorType } from '../shared'
+import {
+  WebSocketReadyState,
+  RippledMethod,
+  WebSocketRequestOptions,
+  WebSocketResponse,
+  WebSocketStatusResponse,
+  WebSocketStatusErrorResponse,
+  WebSocketTransactionResponse,
+} from '../shared/rippled-web-socket-schema'
+
+function sleep(ms: number): Promise<void> {
+  return new Promise((resolve) => setTimeout(resolve, ms))
+}
+
+/**
+ * A network client for interacting with the rippled WebSocket API.
+ * @see https://xrpl.org/rippled-api.html
+ */
+export default class WebSocketNetworkClient {
+  private readonly socket: WebSocket
+  private accountCallbacks: Map<
+    string,
+    (data: WebSocketTransactionResponse) => void
+  >
+  private messageCallbacks: Map<string, (data: WebSocketResponse) => void>
+  private waiting: Map<string, WebSocketResponse | undefined>
+
+  /**
+   * Create a new WebSocketNetworkClient.
+   *
+   * @param webSocketUrl The URL of the rippled node to query.
+   * @see https://xrpl.org/monitor-incoming-payments-with-websocket.html
+   */
+  public constructor(
+    webSocketUrl: string,
+    private handleErrorMessage: (message: string) => void,
+  ) {
+    this.socket = new WebSocket(webSocketUrl)
+    this.accountCallbacks = new Map()
+    this.waiting = new Map()
+
+    this.messageCallbacks = new Map()
+    this.messageCallbacks.set('response', (data: WebSocketResponse) => {
+      const dataStatusResponse = data as WebSocketStatusResponse
+      this.waiting.set(dataStatusResponse.id, data)
+    })
+    this.messageCallbacks.set('transaction', this.handleTransaction)
+
+    this.socket.addEventListener('message', (event) => {
+      const parsedData = JSON.parse(event.data) as WebSocketResponse
+      const dataType = parsedData.type
```

Comment:
can we call this `messageType`? seems like this is the type of the message associated with the payload